jsregexp.c

```diff
@@ -100,6 +113,66 @@ static inline uint16_t *utf8_to_utf16(
 }
 
 
+static int jsstring_new(lua_State* lstate) {
+  size_t input_len;
+  const uint8_t* input = (uint8_t*)luaL_checklstring(lstate, 1, &input_len);
+  struct jsstring* ud;
+  if (utf8_contains_non_ascii((char *) input)) {
+    uint32_t *indices;
+    uint32_t input_utf16_len;
+    uint16_t *input_utf16 = utf8_to_utf16(input, input_len, &input_utf16_len, &indices);
+    
+    if (!input_utf16) {
+      luaL_error(lstate, "malformed unicode");
+    }
+
+    ud = lua_newuserdata(lstate, sizeof(*ud));
+    ud->is_wide_char = true;
+    ud->len = input_utf16_len;
+    ud->u.str16 = input_utf16;
+    ud->bstr = strdup((char*)input);
+    ud->indices = indices;
+  } else {
+    ud = lua_newuserdata(lstate, sizeof(*ud));
+    ud->is_wide_char = false;
+    ud->len = input_len;
+    ud->u.str8 =(uint8_t*) strdup((char*)input);
+    ud->bstr = (char*)ud->u.str8;
+    ud->indices = NULL;
+  }
+  luaL_getmetatable(lstate, JSSTRING_MT);
+  lua_setmetatable(lstate, -2);
+  return 1;
+}
+
+static int jsstring_gc(lua_State* lstate) {
+  struct jsstring *s = lua_touserdata(lstate, 1);
+  free(s->u.str8);
+  free(s->indices);
+  if (s->is_wide_char) {
+    free(s->bstr);
+  }
+  return 0;
+}
+
+static struct luaL_Reg jsstring_meta[] = {
+  {"__gc", jsstring_gc},
+  {NULL, NULL}
+};
+
+static inline struct jsstring* lua_tojsstring(lua_State *lstate, int arg) {
+  if (lua_isuserdata(lstate, arg)) {
+    // already jsstring
+    return (struct jsstring*) luaL_checkudata(lstate, arg, JSSTRING_MT);
+  } else {
+    // coerce to jsstring
+    lua_pushcfunction(lstate, jsstring_new);
+    lua_insert(lstate, arg);
+    lua_call(lstate, 1, 1);
+    return (struct jsstring*) luaL_checkudata(lstate, arg, JSSTRING_MT); 
+  }
+}
+
 static int regexp_call(lua_State *lstate)
```

Comment:
I would honestly just like to deprecate this at some point since it is essentially `reg.match` with a different result layout. Don't put too much effort into updating it (as long as it doesn't break, of course).

Reply:
Ok sounds good